### DISCUSSION TITLE
How do you move -h|--help and -v|--version in the help text?

### DISCUSSION THREAD

**Discussion by samuel-lucas6** [spectreconsole/spectre.console#1468] [Q&A] [answered]
By default, `-h, --help` and `-v, --version` appear at the top of the help text. However, I'd like them at the bottom since the user already knows how to use `--help` to see the help text and `--version` will barely be run compared to other options.

Is there any way to move these to the bottom or hide `-h, --help` completely?

**Comment by patriksvensson** [comment, marked as answer]
For this you will need to implement a custom help provider: https://spectreconsole.net/cli/command-help

**Reply by samuel-lucas6** [reply]
Thanks, I'll have a play around with that. I did wonder if that was the only way.

**Comment by FrankRay78** [comment]
How did you get on @samuel-lucas6 ?  I gave it a go myself, but unfortunately I think it's going to be a rather ugly custom help provider implementation.

The example [CustomHelpProvider](https://github.com/spectreconsole/spectre.console/blob/main/examples/Cli/Help/CustomHelpProvider.cs) demonstrates overriding one of the write functions, namely:

`public override IEnumerable<IRenderable> GetHeader(ICommandModel model, ICommandInfo? command)`

I thought I could follow suit, overring the appropriate method to simply reorder the options to have `-v` and `-h` last. 

The spectre.console [HelpProvider](https://github.com/spectreconsole/spectre.console/blob/main/src/Spectre.Console.Cli/Help/HelpProvider.cs)  renders the options in the following method, which can also be overriden:

```csharp
    /// <summary>
    /// Gets the options section of the help information.
    /// </summary>
    public virtual IEnumerable<IRenderable> GetOptions(ICommandModel model, ICommandInfo? command)
    {
        // Collect all options into a single structure.
        var parameters = HelpOption.Get(command, resources);
```

However, `HelpOption` is a private sealed class, and it's that which includes `-v` and `-h`, see:

```csharp
        public static IReadOnlyList<HelpOption> Get(ICommandInfo? command, HelpProviderResources resources)
        {
            var parameters = new List<HelpOption>();
            parameters.Add(new HelpOption("h", "help", null, null, resources.PrintHelpDescription, null));

            // Version information applies to the entire application
            // Include the "-v" option in the help when at the root of the command line application
            // Don't allow the "-v" option if users have specified one or more sub-commands
            if ((command == null || command?.Parent == null) && !(command?.IsBranch ?? false))
            {
                parameters.Add(new HelpOption("v", "version", null, null, resources.PrintVersionDescription, null));
            }
```

I think all of `HelpProvider.GetOptions` and `HelpOption.Get` would need to be reimplemented in a custom help provider, and I'm not entirely sure that is possible given a number of internal spectre.console classes involved.

Failing that, I guess two other solutions seem possible, namely:

1. `HelpOption` (and equally `HelpArgument` for good measure) are made virtual/overridable, giving you full control over which parameters are included in the rendering, and also their ordering, and/or
2. Some kind of ordering attribute applied to the display of options, with `-v` and `-h` defaulting to first and second as default behaviour (to ensure backward compatibility). [nb. probably my least favoured option]

Truthfully, this behaviour is probably somewhere not very high on the priority list.  

But equally, I'm open-minded and I'd be interested in hearing your thoughts about the above, and also the importance (or not) of this behaviour for your users. Also comments from any other spectre.console users who may have had similar thoughts.

**Reply by samuel-lucas6** [reply]
> How did you get on @samuel-lucas6 ? I gave it a go myself, but unfortunately I think it's going to be a rather ugly custom help provider implementation.

I haven't attempted it so far as I'm not convinced it's worth the hassle. Your analysis is only putting me off more.

[CommandLineUtils](https://github.com/natemcmaster/CommandLineUtils) has `[HelpOption("-h|--help", ShowInHelpText = false)]`, which is the sort of thing I was hoping for. Something easily adjustable.

> Truthfully, this behaviour is probably somewhere not very high on the priority list.

Yeah, it's not a dealbreaker, but it does seem odd to have `-h` and `-v` prioritised when `-h` is required to view the help text, `-v` will probably barely be run compared to other options, and `-h|--help`/`-v|--version` are almost universal among CLI programs. I don't see a good argument for that being the default.

**Reply by FrankRay78** [reply]
I ask because I will probably batch up a few of the CLI issues to address in one go, sometime in the coming months. I think your rationale is valid, I'm not sure the original reason for ordering, certainly possible it never had explicit thought. I'll likely convert this to an issue, in part because I want to move to popularity based prioritisation (ie. Thumbs up counts) in the issue list.

**Reply by FrankRay78** [reply]
Issue created: https://github.com/spectreconsole/spectre.console/issues/1494
